### PR TITLE
adding support for copyfile step

### PR DESCRIPTION
**Pull request by TrevorBoone**

### CODE REVIEW

File: ospackage/recipes/buildcommand.go [hunk 1]
```diff
@@ -46,10 +49,65 @@ func BuildCommand(step *osconfigpb.SoftwareRecipe_Step, artifacts map[string]str
...
+
 	return nil, nil
 }
 
+func parsePermissions(s string) (os.FileMode, error) {
+	if s == "" {
+		return 755, nil
+	}
+
+	i, err := strconv.Atoi(s)
```
[the remark this answers is not in the document]
Comment (TrevorBoone) [reply ospackage/recipes/buildcommand.go]: you're right, this should definitely be octal

File: ospackage/recipes/buildcommand.go [hunk 2]
```diff
@@ -46,10 +49,65 @@ func BuildCommand(step *osconfigpb.SoftwareRecipe_Step, artifacts map[string]str
...
+	}
+
+	if _, err := os.Stat(dest); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		// file exists
+		if !step.FileCopy.Overwrite {
+			logger.Infof("skipping FileCopy step as file at %s already exists", dest)
```
[the remark this answers is not in the document]
Comment (TrevorBoone) [reply ospackage/recipes/buildcommand.go]: I think that this matches the current proto description.

[the remark this answers is not in the document]
Comment (hopkiw) [reply ospackage/recipes/buildcommand.go]: i see. i think the proto should be updated, then. it doesn't make sense to succeed in that case, and it's my fault for not catching the comment detail.

to give an example, if a user has a recipe that first copies a file then runs a command which may involve that file, they will have an undefined behavior if we silently pass because the destination exists. if people want to add logic around their copies, they can use script steps.

[the remark this answers is not in the document]
Comment (TrevorBoone) [reply ospackage/recipes/buildcommand.go]: okay. I'll change it to return an error then

File: ospackage/recipes/buildcommand.go [hunk 3]
```diff
@@ -46,10 +48,64 @@ func BuildCommand(step *osconfigpb.SoftwareRecipe_Step, artifacts map[string]str
...
+		return nil, err
+	}
+
+	permissions, err := parsePermissions(step.FileCopy.Permissions)
+	if err != nil {
+		return nil, err
+	}
+
+	if _, err := os.Stat(dest); err != nil {
+		if !os.IsNotExist(err) {
```
Comment (ryanwe) [inline ospackage/recipes/buildcommand.go]: minor: double negatives are hard to parse, can this be changed to `if os.IsExist(err)`?

File: ospackage/recipes/buildcommand.go [hunk 4]
```diff
@@ -46,10 +48,64 @@ func BuildCommand(step *osconfigpb.SoftwareRecipe_Step, artifacts map[string]str
...
+		// file exists
+		if !step.FileCopy.Overwrite {
+			return nil, fmt.Errorf("in FileCopy step file already exists at path %q and Overwrite = false", step.FileCopy.Destination)
+		}
+		os.Chmod(dest, permissions)
+	}
+
+	src, ok := artifacts[step.FileCopy.ArtifactId]
+	if !ok {
+		return nil, fmt.Errorf("Could not find location for artifact %q in FileCopy step", step.FileCopy.ArtifactId)
```
Comment (ryanwe) [inline ospackage/recipes/buildcommand.go]: The previous error starts with `in FileCopy step`, should this one also start with that or is there a reason to be inconsistent?

File: ospackage/recipes/buildcommand.go (hunk 3 above)
Comment (TrevorBoone) [reply ospackage/recipes/buildcommand.go]: os.IsNotExist is oddly not the opposite of os.IsExist actually (despite the names seeming that way). Each of those only match a very specific set of errors (I think this is the relevant source: https://golang.org/src/os/error_unix.go)

File: ospackage/recipes/buildcommand.go (hunk 4 above)
Comment (TrevorBoone) [reply ospackage/recipes/buildcommand.go]: I took out "in FileCopy step" and changed installrecipe to wrap the error with the step number to make debugging easier

**Comment by hopkiw** [comment]
@TrevorBoone please resolve conflicts for merge
